src/main/java/fi/aalto/cs/apluscourses/presentation/MainViewModel.java

```diff
@@ -115,12 +119,22 @@ public void setAuthentication(Authentication auth) {
   public void readAuthenticationFromStorage(@Nullable PasswordStorage passwordStorage,
                                             @NotNull TokenAuthentication.Factory factory) {
     if (hasTriedToReadAuthenticationFromStorage.getAndSet(true) || authentication.get() != null) {
+      exercisesViewModel.get().setProjectReady(true);
+      //courseViewModel.get().setProjectReady(true);
       return;
     }
     Optional.ofNullable(passwordStorage)
         .map(PasswordStorage::restorePassword)
         .map(factory::create)
         .ifPresent(this::setAuthentication);
+    exercisesViewModel.get().setProjectReady(true);
+    //courseViewModel.get().setProjectReady(true);
+    if (!exercisesViewModel.get().isAuthenticated()) {
+      this.updateExercises(); // When authentication is null mainViewModel is not updated?
```

Comment:
> When authentication is null mainViewModel is not updated?

Correct. Calling `setAuthentication` a few lines above causes `updateExercise` to be executed*. However, it's only called when the authentication read from the password storage is not null (note `ifPresent`).

---

*) Because of this line in the constructor:
```java
authentication.addValueObserver(this, MainViewModel::updateExercises);
```